type-generator/src/main/scala/mesosphere/raml/ir/ConstraintT.scala

```diff
@@ -0,0 +1,156 @@
+package mesosphere.raml.ir
+
+import mesosphere.raml.backend.{PlayReads, camelify, scalaFieldName, underscoreToCamel}
+import treehugger.forest._
+import definitions._
+import treehugger.forest
+import treehuggerDSL._
+
+import scala.annotation.tailrec
+
+sealed trait ConstraintT[C] { self =>
+  val name: String
+  val constraint: C
+  val constraintToValue: C => Tree = { (c: C) => LIT(c) } // decent assumption for built-ins, probably not much else
+
+  /** false indicates custom, non-play constraint implementations that we've implemented as part of this generator */
+  val builtIn: Boolean
+
+  /** @return a code gen expression that represents a playJS reads validation */
+  def validate(): Tree
+
+  /** a code gen expression for a field that represents the constraint limit */
+  val limitField: Option[Tree] = None
+
+  /** decorate this constraint with a `limitField` implementation */
+  def copyWith(lf: Option[Tree] = None): ConstraintT[C]
+
+  def withFieldLimit(f: FieldT): ConstraintT[C] = {
+    val fieldName = scalaFieldName(underscoreToCamel(camelify(s"constraint_${f.rawName}_${name}".replace("-", "_"))))
+    val limit = (VAL(fieldName) := constraintToValue(constraint))
+    copyWith(Option(limit))
+  }
+}
+
+object ConstraintT {
+
+  // built-in playJS validators
+
+  case class MaxLength(len: Integer, override val limitField: Option[forest.Tree] = None) extends BasicConstraint[Integer] {
```

Comment:
Compare this to the old `def MaxLength` in `type-generator/src/main/scala/mesosphere/raml/ir/Constraint.scala`